repair/repair.cc

```diff
@@ -1612,3 +1612,96 @@ future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_me
+                std::unordered_set<inet_address> neighbors_set(new_eps.begin(), new_eps.end());
+                bool skip_this_range = false;
+                // For decommission operation, the decommission node will
+                // repair all the ranges the leaving node is responsible for.
+                if (is_removenode) {
+                    // For removenode operation, the new owner of the range
+                    // will repair the data. Nodes that are not the new owner
+                    // of a range will ignore the range. There can be only one
+                    // new owner for each of the ranges.
+                    // Note: The new owner is the node that does not own the
+                    // range before but owns the range after the remove
+                    // operation. It does not have to be the primary replica of
+                    // the range.
+                    auto new_owner = neighbors_set;
+                    for (const auto& node : current_eps) {
+                        new_owner.erase(node);
+                    }
+                    if (new_owner.size() != 1) {
+                        rlogger.warn("{}: keyspace={}, range={}, current_replica_endpoints={}, new_replica_endpoints={}",
+                            op, keyspace_name, r, current_eps, new_eps);
+                        throw std::runtime_error("Can not find new owner node!");
+                    }
+                    skip_this_range = *new_owner.begin() != myip;
+                }
+                // Calculate the neighbors
+                neighbors_set.insert(current_eps.begin(), current_eps.end());
+                neighbors_set.erase(myip);
+                neighbors_set.erase(leaving_node);
+                std::vector<inet_address> neighbors(neighbors_set.begin(), neighbors_set.end());
+
+                if (skip_this_range) {
+                    // Let the new owner node to fix this range, other nodes skip this range
+                    rlogger.debug("{}: keyspace={}, range={}, current_replica_endpoints={}, new_replica_endpoints={}, neighbors={}, skipped",
+                        op, keyspace_name, r, current_eps, new_eps, neighbors);
+                } else {
+                    // The decommission node will fix the range
+                    rlogger.debug("{}: keyspace={}, range={}, current_replica_endpoints={}, new_replica_endpoints={}, neighbors={}",
+                        op, keyspace_name, r, current_eps, new_eps, neighbors);
+                    range_sources[r] = repair_neighbors(std::move(neighbors));
+                    if (is_removenode) {
+                        ranges_for_removenode.push_back(r);
+                    }
+                }
+            }
+            if (is_removenode) {
+                ranges.swap(ranges_for_removenode);
+            }
+            auto nr_ranges = ranges.size();
+            sync_data_with_repair(db, keyspace_name, std::move(ranges), std::move(range_sources)).get();
```

Comment:
You see, here for instance `ranges` and `range_sources` are going to be pretty big. 
This relates to my other comment about the `sync_data_with_repair()` where I wrote that it would make sense try to not duplicate them across shards.

Reply:
Like we discussed in the other comment thread, repair_neighbors is only of size RF not the whole number of nodes in the cluster. I see your point to share between shards. However it adds addiction complexity for the cross shard sharing between multiple shard local repair_info objects. Perhaps we can do it later, at least not this iteration. If I may, I would mark this conversation as resolved.

Reply:
Sure. Please do (mark it as resolved).
It's just a nitpick.

Reply:
Thanks